### PR TITLE
Error out on non-private calls to private methods

### PR DESCRIPTION
**Pull request by bss**
This will type-check private methods and ensure they can actually be called. Based off of https://github.com/sorbet/sorbet/compare/jez-private-methods

The following can still not be type-checked:
- splat/splat&block/block methods since they do not carry visibility information when being switched to the magic replacements.
- "subsequent visibility": using the private keyword, thus marking the following methods private.

I intend to work on those separately though.

### Motivation
Currently sorbet does not warn users about calling a private method on an object, instead a runtime error will occur. This will make sorbet error out  instead.

### Test plan
See included automated tests.

### CODE REVIEW

File: infer/environment.cc
Comment (bss) [inline infer/environment.cc:837]: Unfortunately we need to check that the method is a method here. I didn't spend a lot of time debugging, but from initial debugging it looked like `it->main.method.exists()` is false when initializing an object (e.g. `Object.new`). I think we'll need to fix this separately though and remove the check here when that's fixed.

Comment (jez) [reply infer/environment.cc:837]: ok i'd be more comfortable fixing that first and landing private methods second

File: test/testdata/infer/private_methods.rb
Comment (jez) [inline test/testdata/infer/private_methods.rb:62]: Can you leave the TODOs in the code where the changes would be required?

Also, I don't think it would be hard to support these off the bat, would it? Can you similarly pass along the flags when synthesizing a `<Magic>.<callWithSplat>` method call?

File: test/testdata/infer/private_methods.rb
Comment (jez) [inline test/testdata/infer/private_methods.rb:66]: Huh. It used to, AFAIK. Something must have changed.

Stripe bans this form of declaring things private, so I'm actually ok leaving this as a missing feature and either fixing it after landing this PR or making an issue and leaving it for whoever wants its the most.

File: test/testdata/infer/private_methods.rb
Comment (jez) [reply test/testdata/infer/private_methods.rb:1]: Methods that are defined at the top-level of a file are implicitly marked private:

```ruby
def foo; end
def self.bar; end
```

^ these are both private, and both defined as instance methods of `Object`. Can you double check that `Object.new.foo` and `Object.new.bar` both error?

File: infer/environment.cc
Comment (bss) [reply infer/environment.cc:837]: That's fine. Let me take a stab a figuring out what's going on.

File: test/testdata/infer/private_methods.rb
Comment (jez) [reply test/testdata/infer/private_methods.rb:1]: Can you add a `symbol-table.exp` file for this test?

You should be able to run

```
touch test/testdata/infer/private_methods.rb.symbol-table.exp
tools/scripts/update_testdata_exp.sh
```

to add the new expectation file and record it

that should show us that all these methods' symbols have the `private` flag set, and will help to debug if this test ever starts failing.

**Review by jez — COMMENTED** [review]
This is looking really good.

It's at a point where I'm already curious to see how this fares on Stripe's codebase, so I'll probably do that soon.

There are a couple small things, and a couple bigger things that I'd like to hear your thoughts on.

File: test/testdata/resolver/sig_generated.rb
Comment (bss) [reply test/testdata/resolver/sig_generated.rb:7]: Honestly, I was surprised too 😂

File: test/testdata/infer/private_methods.rb
Comment (bss) [reply test/testdata/infer/private_methods.rb:66]: I'll do it in a subsequent PR as this whole thing is useless to us at Gusto if it doesn't handle this correctly (it's almost exclusively how we mark things private). But figured it was separate enough from this PR that I wanted to do it separately. So expect something incoming soon-ish.

File: test/testdata/infer/private_methods.rb
Comment (bss) [reply test/testdata/infer/private_methods.rb:62]: I can take a quick stab before declaring to wait. Figured the probability of hitting a method marked both private and being a splat and/or block method was on the lower side, so it could wait.

File: test/testdata/infer/private_methods.rb
Comment (bss) [reply test/testdata/infer/private_methods.rb:1]: Great ideas for improving the tests, I'll do that.

File: infer/environment.cc
Comment (bss) [reply infer/environment.cc:837]: @jez 
So, I've found the issue.
It's pretty easy to reproduce by removing the check for `exists()` and `isMethod()` in this line and then running sorbet on a file containing:
```
# typed: true
Object.new
```
It happens when `#initialize` is not overloaded for a class.

Tracing through the code `main.method.exists()` becomes false during the dispatch to `#initialize` from `.new`. In that call:

https://github.com/sorbet/sorbet/blob/master/core/types/calls.cc#L1173-L1181

`main` is replaced with the `main` for `#initialize`. `#initialize` is special cased in `dispatchCallSymbol`:

https://github.com/sorbet/sorbet/blob/master/core/types/calls.cc#L497-L511

That code sets the `SymbolRef` for the returned `DispatchResult` to be `Symbols::noSymbol()`, which will return `false` on `exists()`.

Now the question is how we fix that. Could we create a new static symbol ref, e.g.:
```
static SymbolRef DefaultInitializeMethod() {
  return SymbolRef(nullptr, SymbolRef::Kind::Method, XXXX); // TODO: figure out the right id that is != 0
}
```

and use that instead of `Symbols::noSymbol()` ? Since the _id would be `!= 0`, `exists()` should return `true`, and since it has `SymbolRef::Kind::Method`, `isMethod` should return `true` as well. Do you see any downside to this approach?

Comment (bss) [reply infer/environment.cc:837]: Just tried this and that doesn't work. It seems to be due to sorbet not being able to lookup the `Symbol` from the `SymbolRef`. I looked into adding an initialize method as `GlobalState.cc`, but that made me think about the comment above about not being able to define it on `BasicObject` and having to special case it. Is it still the case that that is needed? Do you know what the special cases is where that would shadow the real definition? It seem like the right solution to me. But perhaps you can think of something simpler?

Comment (jez) [reply infer/environment.cc:837]: The thing I’m more worried about is not that a symbol doesn’t exist (we might be able to work around some stuff with initialize)

But I’m very concerned that if the dispatch main exists that it’s not a method symbol

File: test/testdata/infer/private_methods.rb
Comment (Morriar) [inline test/testdata/infer/private_methods.rb:30]: There are also private class methods defined like this:

```rb
class Foo
  class << self
    private def foo; end
  end
end

Foo.foo # error: Non-private call to private method `Foo#foo`
```

**Comment by Morriar** [comment]
Awesome feature!

> It's at a point where I'm already curious to see how this fares on Stripe's codebase, so I'll probably do that soon.

I'm eager to try this on our codebase too.

As to make the transition easier for large codebases, could we enable it with a `--enforce-private` flag at first?

File: test/testdata/infer/private_methods.rb
Comment (bss) [reply test/testdata/infer/private_methods.rb:30]: Good call. Even though I would expect it to just work, we do need a bunch of tests for class methods as well

**Comment by bss** [comment]
@Morriar I've been going back and forth on that. The arguments talking for a special flag is that the check will either:
1. Crash/report an error that isn't actually an error.
2. Make sorbet error out due to an actual error in code.

As for 1., the constructs in sorbet being used are to few and it's just an additional error check that I am not super worried about it. This could very well be where I am wrong, but running it across the Stripe, Shopify and Gusto codebases should give quite the coverage IMO (at least we use private instance and class methods extensively).

If there is an actual error in the code, I think it should be fine to error on it when running sorbet, the alternative is that the ruby runtime will when the code is run.

A 3rd option of this check being less strict than the ruby runtime I am less worried about. That's something we can fix along the way.

Any thoughts?

File: infer/environment.cc
Comment (bss) [reply infer/environment.cc:837]: @jez I don't entirely follow what you mean by your last sentence, can you expand a bit on that?

Comment (bss) [reply infer/environment.cc:837]: Ah I see. Let me double check but pretty sure all the cases I've seen so far only needed the `.exists()` guard, agreed it would be worry-some if it was suddenly not a method in here.

**Comment by jez** [comment]
> As to make the transition easier for large codebases, could we enable it with a `--enforce-private` flag at first?

I'm happy to be proven wrong, but I'm guessing that that most existing code does not have private method call errors.

An option we have would be to add an autocorrect that rewrites `foo.bad_private_call(` to `foo.send(:bad_private_calll,` so that we could silence all the errors and fix them async.

We don't have any other errors enabled or disabled behind a flag in sorbet, and this has been a very intentional design decision (all errors are reported or silenced with strictness levels).

If you really don't want to see the private methods errors, you can `--error-black-list=7031` in your `sorbet/config` file.

**Comment by jez** [comment]
@bss Can you add something to the error message as context like "the method was defined here"

Here's an example to cargo cult:

https://sorbet.run/#%23%20typed%3A%20true%0A%0Adef%20foo%3B%20end%0Adef%20foo(x)%3B%20end

Also I only see 6 new errors from this PR on Stripe's codebase. I'm going to look into whether they're bugs in Stripe's codebase or this PR.

**Comment by jez** [comment]
The breakdown of the 6 errors:

- 3 times: someone subclassed a class with a private method `foo` and called `public :foo` so they could call it in a test. i think sorbet should reject this, as it does in this PR. workarounds: `def foo; super; end`, `.send` at the call site
- 1 times: someone is calling a private method wrapped in `assert_raises(Exception)` that test is garbage; they were hoping that something else raised an exception, not a type error for the private method call
- 1 times: someone is calling a private method from a db migration script dated 2020-03-11. It is probably not tested, and was probably made private after the migration ran.
- 1 times: call hidden behind `foo&.bar`, and I think that either the code never runs or `foo` is always `nil`

File: test/testdata/infer/private_methods.rb
Comment (bss) [reply test/testdata/infer/private_methods.rb:62]: Took another look at this. I do pass on the flags of `isPrivateOk` when for `MK::Send`, so that part is covered. What's missing is to know if the method is private or not.
Please correct me if I misunderstood the built-in magic methods like `call-with-block`, but as far as I understand there only exist a single method for `call-with-block`, no matter how many times or on which methods it's used:

https://github.com/sorbet/sorbet/blob/7b333e9c50a020924503b0abf8712ee118f8bae1/core/GlobalState.cc#L519-L530

So I don't think I can set the private method flag (`setMethodPrivate`) on that without doing so across all usages of it?

Perhaps we could split the `callWithBlock` magic method into two methods: `callWithBlockPublic` & `callWithBlockPrivate`. But that doesn't feel right design-wise to me.

Comment (jez) [reply test/testdata/infer/private_methods.rb:66]: You're right, that's not the right approach. I thought it would be as simple as setting the flags on the constructed send in Desugar.cc but that wasn't enough. I'm fine not blocking on that to land this.

**Comment by bss** [comment]
@jez This is ready for another look. As we chatted about on slack, changing the `super` calls and dispatches on untyped receivers to use `Symbols::noSymbol` instead of `Symbols::untyped` as you suggested allowed me to remove the `isMethod` guard.
I've gone over the rest of the comments and addressed them, with the exception of:
1. Handling visibility of subsequent methods.
2. Handling the splat/block magic functions.

As I would prefer to get this merged and do it in subsequent PRs.

**Comment by jez** [comment]
> I've gone over the rest of the comments and addressed them, with the exception of:
> 
> 1. Handling visibility of subsequent methods.
> 2. Handling the splat/block magic functions.
> 
> As I would prefer to get this merged and do it in subsequent PRs.

👍 this works for me!

**Review by jez — APPROVED** [review]
Awesome!

I will work on fixing the problems with Stripe's codebase today and then we can merge this.

**Pull request merged**
